### ISSUE TITLE
Typing for Filtered Iterables

### ISSUE THREAD

**Issue by cSDes1gn** [microsoft/pyright#2073]
I would expect type hints would be evaluated for iterables which are filtered by some conditional function:
```python
li: List[Optional[Union[str]]] = ["hello", "test", None]
filtered_li = list(filter(lambda el: el is not None, li)) # type: List[str | None]
```

Specifically, because list comprehension provides correct type hinting:
```python
filtered_li = [el for el in li if el is not None] # type: List[str]
```

I am suggesting we adopt a similar behavior for filter functions as is done for list comprehension. Open to discussion regarding the implementation semantics.

**Comment by JelleZijlstra** [comment]
This is doable with PEP 647 TypeGuards (https://www.python.org/dev/peps/pep-0647/). This should be done in typeshed, and there's already an issue open for it: python/typeshed#5661. We'd appreciate a PR adding the implementation.

**Comment by erictraut** [comment]
A type checker like pyright knows about the built-in semantics of a list comprehension, but it doesn't know about the semantics or behaviors performed by functions like `filter`. As @JelleZijlstra mentioned, these semantics would need to be codified in a `TypeGuard` directive within the typeshed stubs.
